### PR TITLE
Changed parameter of ShouldlyAssertionContext to IShouldlyAssertionConte...

### PR DESCRIPTION
**Pull request by JoeMighty**
Changed parameter of `ShouldlyAssertionContext` to `IShouldlyAssertionContext` on `ShouldlyMessage.CreateActualVsExpectedMessage()`.

Also converted private method `ShouldlyMessage.CreateActualVsExpectedMessage()` to be a static method.

### CODE REVIEW

**Comment by JakeGinnivan** [comment]
Should the private field and protected property be make IShouldlyAssertionContext as well?

Also, do we even need the interface?

**Comment by JoeMighty** [comment]
That's a good point, they could also be made `IShouldlyAssertionContext` too.

As for whether the interface is needed, `ShouldlyAssertionContext` and a test equivalent are the only implementations of it the interface. Are you thinking of replacing the tests with a concrete type of `ShouldlyAssertionContext`?

**Comment by JakeGinnivan** [comment]
If the interface is used for testing, then let's leave the interface. If not remove it.

If we keep it lets use it everywhere.

**Comment by JoeMighty** [comment]
Sounds good. I've updated the references of `ShouldlyAssertionContext` to `IShouldlyAssertionContext`.

File: src/Shouldly/ShouldlyMessage.cs [hunk 1]
```diff
@@ -76,12 +76,12 @@ internal abstract class ShouldlyMessage
...
         };
-        private ShouldlyAssertionContext _shouldlyAssertionContext;
+        private IShouldlyAssertionContext _shouldlyAssertionContext;
 
-        protected ShouldlyAssertionContext ShouldlyAssertionContext
+        protected IShouldlyAssertionContext ShouldlyAssertionContext
         {
             get { return _shouldlyAssertionContext; }
-            set { _shouldlyAssertionContext = value; }
+            set { _shouldlyAssertionContext = value.As<IShouldlyAssertionContext>(); }
```
Comment (JakeGinnivan) [inline src/Shouldly/ShouldlyMessage.cs]: Do we need this `.As<IShouldlyAssertionContext>()`?

Comment (JoeMighty) [inline src/Shouldly/ShouldlyMessage.cs]: No, this is no longer needed. I've just removed it.

**Comment by JakeGinnivan** [comment]
With pull requests, im a fan of rebasing. It is an awesome git feature.

If you:

```
git fetch upstream
git rebase upstream/master
```

It will remove those merge commits and rewrite history so your commits were done *after* the changes to master. This visualises it quite well http://onlywei.github.io/explain-git-with-d3/#rebase

Another tip is to work on feature branches. So whenever you want to start a pull request you go through this workflow

```
git checkout master
git fetch upstream
git rebase upstream/master # you can also 'merge --ff-only` which only allows a fast forward merge. 
                                            # Rebase means if you accidently commited to master at some point your change will just be moved to be the newest commit
git checkout -b feature/MyFeature  # you can name your branch anything
# work normally, commit etc
git push origin feature/MyFeature
#then create a pull request from feature/MyFeature 
# That means you can have multiple feature branches at any one time
```

Hope that helps :)

If you could rebase this PR it will remove those two merge commits and keep the git history cleaner.

**Comment by JoeMighty** [comment]
Thank you for your patience Jake; this is the first time I've worked collaboratively on a project using Git.

I've just rebased the PR as per your instructions above; hopefully it's worked as you described. Moving forward I will create a separate branch for any issues I take on or changes I make!

**Comment by JakeGinnivan** [comment]
No problems, I struggled when starting and it is not much effort on my part to help you get into it :) will have a look tomorrow

Sent from my Windows Phone
________________________________

**Comment by JakeGinnivan** [comment]
I think something went slightly wrong with the rebase. There is still a merge commit and two of the commits were duplicated. Have fixed/rebased and merged manually at ee5c5873959c8bfd9ea4610baa5774fe12a16004

Thanks for the PR again. 

If you are interested in learning more about the magic behind git, I found this talk very handy - https://vimeo.com/49444883

**Pull request closed without merging**